### PR TITLE
rtags 2.0 (new formula)

### PR DESCRIPTION
**Pull request by smoofra**
rtags is a ctags-like source code cross-referencer with a clang frontend.

from rtags README:

RTags is a client/server application that indexes C/C++ code and keeps a
persistent in-memory database of references, declarations, definitions,
symbolnames etc. There’s also limited support for ObjC/ObjC++. It allows you to
find symbols by name (including class and namespace scope). Most importantly we
give you proper follow-symbol and find-references support. We also have neat
little things like rename-symbol, integration with clang’s “fixits”
(http://clang.llvm.org/diagnostics.html). We also integrate with flymake using
clang’s vastly superior errors and warnings. Since RTags constantly will reindex
“dirty” files you get live updating of compiler errors and warnings. Since we
already know how to compile your sources we have a way to quickly bring up the
preprocessed output of the current source file in a buffer.

While existing taggers like gnu global, cscope, etags, ctags etc do a good job
for C they often fall a little bit short for C++. With its incredible lexical
complexity, parsing C++ is an incredibly hard task and we make no bones about
the fact that the only reason we are able to improve on the current tools is
because of clang (http://clang.llvm.org/). RTags is named RTags in recognition
of Roberto Raggi on whose C++ parser we intended to base this project but he
assured us clang was the way to go. The name stuck though # Please enter the
commit message for your changes. Lines starting # with '#' will be ignored, and
an empty message aborts the commit.  # On branch rtags2.0 # Changes to be
committed: # new file: Library/Formula/rtags.rb #

### CODE REVIEW

**Comment by dunn** [comment]
Since the resource is for a submodule that's not included with the source code tarball, let's instead have `stable` clone the repository instead:
```ruby
url "https://github.com/Andersbakken/rtags.git",
    :tag => "v2.0",
    :revision => "ba85598841648490e64246be802fc2dcdd45bc3c"
```

**Comment by smoofra** [comment]
ok

**Comment by dunn** [comment]
Audit checks from the bot:
```
        ==> brew style rtags
== /usr/local/Library/Formula/rtags.rb ==
C: 22: 27: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

1 file inspected, 1 offense detected
Error: 2 problems in 1 formula

==> audit problems
rtags:
 * `head` (line 10) should be put before `stable block` (line 4)
 * Formula should have a desc (Description).
```
The stable URL doesn't need to be in a `stable do` block anymore, since there's just the URL.

**Comment by smoofra** [comment]
fixed

File: Library/Formula/rtags.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    ENV.prepend_path "PATH", "#{opt_libexec}/llvm/bin"
+
+    # Homebrew llvm libc++.dylib doesn't correctly reexport libc++abi
+    ENV.append("LDFLAGS", "-lc++abi")
+
+    mkdir "build" do
+      args = std_cmake_args
+      args << ".."
+
+      system "cmake", *args
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: Unless we're passing arguments beyond this, let's just make this `system "cmake", "..", *std_cmake_args`

File: Library/Formula/rtags.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    system "sh", "-c", "rc >/dev/null --help  ; test $? == 1"
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: What is this testing?

File: Library/Formula/rtags.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  depends_on "cmake" => :build
+  depends_on "llvm" => "with-clang"
+  depends_on "openssl"
+
+  def install
+    # we use brew's LLVM instead of the macosx llvm because the macosx one
+    # doesn't include libclang.
+    ENV.prepend_path "PATH", "#{opt_libexec}/llvm/bin"
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: This isn't pointing at the `LLVM` keg, with this line you're adding the libexec path of the `rtags` Cellar to `$PATH`.

File: Library/Formula/rtags.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+class Rtags < Formula
+  desc "ctags-like source code cross-referencer with a clang frontend"
+  homepage "https://github.com/Andersbakken/rtags"
+
+  head "https://github.com/Andersbakken/rtags.git"
+
+  stable do
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: Don't need a `stable do` block unless you're doing something that only applies to stable, like a patch or resource.

File: Library/Formula/rtags.rb (hunk 2 above)
Comment (smoofra) [inline Library/Formula/rtags.rb]: its just testing that rc --help can be run.

**Comment by smoofra** [comment]
@DomT4  fixed those things you mentioned

File: Library/Formula/rtags.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+class Rtags < Formula
+  desc "ctags-like source code cross-referencer with a clang frontend"
+  homepage "https://github.com/Andersbakken/rtags"
+
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: Don't need the empty line here.

File: Library/Formula/rtags.rb (hunk 2 above)
Comment (DomT4) [inline Library/Formula/rtags.rb]: Let's Rubyify it instead.

```ruby
assert_match /some output/, shell_output("#{bin}/rc --help", 1)
```

Or similar.

File: Library/Formula/rtags.rb (hunk 5 above)
Comment (smoofra) [inline Library/Formula/rtags.rb]: ok

**Comment by smoofra** [comment]
bump.

File: Library/Formula/rtags.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    mkdir "build" do
+      system "cmake", "..", *std_cmake_args
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    # not using shell_output because on HEAD the exit code will be 0, but on
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: Did you test this with `brew test --verbose rtags --HEAD`?

File: Library/Formula/rtags.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    # stable it will be 1.
+    cmd = "#{bin}/rc --help"
+    ohai cmd
```
Comment (DomT4) [inline Library/Formula/rtags.rb]: Let's not ohai.

Comment (smoofra) [inline Library/Formula/rtags.rb]: ok

File: Library/Formula/rtags.rb (hunk 6 above)
Comment (smoofra) [inline Library/Formula/rtags.rb]: yes

Comment (DomT4) [inline Library/Formula/rtags.rb]: Can you provide the terminal printout?

Comment (smoofra) [inline Library/Formula/rtags.rb]: 🍺  /usr/local/Cellar/rtags/HEAD: 11 files, 2.7M, built in 39 seconds
lawrence_danna: /usr/local$ brew test --verbose --HEAD rtags
Testing rtags
lawrence_danna: /usr/local$

Comment (DomT4) [inline Library/Formula/rtags.rb]: That seems to have exited okay? There's no failure message there?

Comment (smoofra) [inline Library/Formula/rtags.rb]: yea.. it worked for me.   Is it failing for you?

File: Library/Formula/rtags.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      system "make"
+      system "make", "install"
+    end
+  end
+
+  test do
+    # not using shell_output because on HEAD the exit code will be 0, but on
+    # stable it will be 1.
+    cmd = "#{bin}/rc --help"
+    assert_match /rc options/, `#{cmd}`
```
Comment (mikemcquaid) [inline Library/Formula/rtags.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

Comment (smoofra) [inline Library/Formula/rtags.rb]: ok.

**Comment by smoofra** [comment]
I added a test that actually verifies rtags can do its job.

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by apjanke** [comment]
FYI: Upstream `rtags` has revoked the 2.0 release tag, and is not making releases; see Andersbakken/rtags#582. It's causing the #48462 test to fail. Going to move this to the head-only tap.

**Comment by apjanke** [comment]
Boneyarding (#48528) since head-only tap is deprecated.

**Pull request closed without merging**